src/Symfony/Component/Message/Asynchronous/Middleware/SendMessageMiddleware.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Message\Asynchronous\Middleware;
+
+use Symfony\Component\Message\Asynchronous\Routing\SenderLocatorInterface;
+use Symfony\Component\Message\Asynchronous\Transport\ReceivedMessage;
+use Symfony\Component\Message\MiddlewareInterface;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class SendMessageMiddleware implements MiddlewareInterface
+{
+    private $senderLocator;
+
+    public function __construct(SenderLocatorInterface $senderLocator)
+    {
+        $this->senderLocator = $senderLocator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle($message, callable $next)
+    {
+        if ($message instanceof ReceivedMessage) {
+            return $next($message->getMessage());
+        }
+
+        if (!empty($senders = $this->senderLocator->getSendersForMessage($message))) {
+            foreach ($senders as $sender) {
+                $sender->send($message);
+            }
+
+            if (!in_array(null, $senders, true)) {
+                return;
+            }
```

Comment:
Why would there be a `null`? Shouldn't it be forbidden at the config level?

Reply:
`null` would allow you to rout the message to a sender while still calling the handler locally. See [this note in the `README.md` file](https://github.com/symfony/symfony/pull/24411/files#diff-aa7b060ee5794d3e1e6dd4b34a1c996fR202).